src/core/value/constant.jl

```diff
@@ -106,49 +106,153 @@ identify(::Type{Value}, ::Val{API.LLVMConstantAggregateZeroValueKind}) = Constan
-Base.convert(::Type{Array{T,1}}, ca::ConstantArray) where {T<:Integer} =
-    [convert(T,ConstantInt(ca[i])) for i in 1:length(ca)]
-Base.convert(::Type{Array{T,1}}, ca::ConstantArray) where {T<:AbstractFloat} =
-    [convert(T,ConstantFP(ca[i])) for i in 1:length(ca)]
+function Base.size(ca::ConstantArray)
+    dims = Int[]
+    typ = llvmtype(ca)
+    while typ isa ArrayType
+        push!(dims, length(typ))
+        typ = eltype(typ)
+    end
+    return Tuple(dims)
+end
+Base.length(ca::ConstantArray) = prod(size(ca))
+Base.axes(ca::ConstantArray) = Base.OneTo.(size(ca))
+
+function Base.getindex(ca::ConstantArray, idx::Integer...)
+    # multidimensional arrays are represented by arrays of arrays,
+    # which we need to 'peel back' by looking at the operand sets.
+    # for the final dimension, we use LLVMGetElementAsConstant
+    @boundscheck Base.checkbounds_indices(Base.Bool, axes(ca), idx) ||
+        throw(BoundsError(ca, idx))
+    I = CartesianIndices(size(ca))[idx...]
+    for i in Tuple(I)
+        if isempty(operands(ca))
+            ca = LLVM.Value(API.LLVMGetElementAsConstant(ca, i-1))
+        else
+            ca = (Base.@_propagate_inbounds_meta; operands(ca)[i])
+        end
+    end
+    return ca
+end
+
+# structs
 
 @checked struct ConstantStruct <: ConstantAggregate
     ref::API.LLVMValueRef
 end
 identify(::Type{Value}, ::Val{API.LLVMConstantStructValueKind}) = ConstantStruct
 
+# anonymous
+ConstantStruct(values::Vector{<:Constant}; packed::Core.Bool=false) =
+    ConstantStruct(API.LLVMConstStruct(values, length(values), convert(Bool, packed)))
+ConstantStruct(values::Vector{<:Constant}, ctx::Context; packed::Core.Bool=false) =
+    ConstantStruct(API.LLVMConstStructInContext(ctx, values, length(values), convert(Bool, packed)))
+
+# named
+ConstantStruct(typ::StructType, values::Vector{<:Constant}) =
+    ConstantStruct(API.LLVMConstNamedStruct(typ, values, length(values)))
+
+# create a ConstantStruct from a Julia object
+function ConstantStruct(value::T, ctx::Context=GlobalContext();
+                        anonymous::Core.Bool=false, packed::Core.Bool=false) where {T}
+    isbitstype(T) || throw(ArgumentError("Can only create a ConstantStruct from an isbits struct"))
+    constants = Vector{Constant}()
+    for fieldname in fieldnames(T)
+        field = getfield(value, fieldname)
+
+        if isa(field, Core.Bool)
+            typ = LLVM.Int1Type(ctx)
+            push!(constants, ConstantInt(typ, Int(field)))
+        elseif isa(field, Integer)
+            push!(constants, ConstantInt(field, ctx))
+        elseif isa(field, AbstractFloat)
+            push!(constants, ConstantFP(field, ctx))
+        else # TODO: nested structs?
+            throw(ArgumentError("only structs with boolean, integer and floating point fields are allowed"))
+        end
+    end
+
+    if anonymous
+        ConstantStruct(constants, ctx; packed=packed)
+    else
+        typ = StructType(String(nameof(T)), ctx)
```

Comment:
In the case of named structs we should let the user provide the `StructType` instead of defining one ourselves. This is because a named constant struct is only useful when you have a concrete struct type of which you want to create a constant value.

Also note that `LLVM.StructType("name", ctx) != LLVM.StructType("name", ctx)`, meaning this implementation creates a new (opaque) `StructType` on every call, even for the same `T`.

Reply:
> Also note that `LLVM.StructType("name", ctx) != LLVM.StructType("name", ctx)`, meaning this implementation creates a new (opaque) `StructType` on every call, even for the same `T`.

Most of the time, this code is used in `llvmcall` context where we get a new module without any pre-existing types, so this shorthand is still useful I think. But a version that accepts a StructType is useful too.

Reply:
Alternative:

```
julia> constant_struct = ConstantStruct(test_struct)
%TestStruct { i1 false, i64 52, half 0xHC100 }

julia> constant_struct = ConstantStruct(test_struct)
ERROR: ArgumentError: Type name 'TestStruct' is already used in this context. Use a different context, name, or request an anonymous struct.
Stacktrace:
 [1] ConstantStruct(value::TestStruct, ctx::Context; name::String, anonymous::Bool, packed::Bool)
   @ LLVM ~/Julia/pkg/LLVM/src/core/value/constant.jl:232
 [2] ConstantStruct
   @ ~/Julia/pkg/LLVM/src/core/value/constant.jl:211 [inlined]
 [3] top-level scope
   @ REPL[5]:1

julia> constant_struct = ConstantStruct(test_struct; name="foo")
%foo { i1 false, i64 52, half 0xHC100 }

julia> constant_struct = ConstantStruct(test_struct; anonymous=true)
{ i1, i64, half } { i1 false, i64 52, half 0xHC100 }
```

Reply:
And the explicit constructor taking a StructType is still there, just not for the convenience constructor taking Julia values.